accepted/2021/experimental-package-shading/experimental-package-shading.md

```diff
@@ -0,0 +1,256 @@
+
+Lastly, consider a case where the package project has a shaded reference to a package that's also a dependency of an unshaded reference.
+
+![block9](experimental-package-shading.md.9.png)
+
+This is **explicitly disallowed** at this time for the same reason as the previous example: the package project's direct and transitive references to `Bar` must unify. The only way to do this is by shading `Foo` so that its references to `Bar` can be updated to target the shaded version, and we require the developer do this explicitly so that all shading operations are clear and intentional.
+
+## FAQ
+
+### Why not runtime side-by-side assembly loading?
+
+By extending the runtime to support side-by-side loading of different versions of the same assembly, we would remove the need to manipulate assemblies and their bytecode. However, the other parts of this proposal would still be needed - specifying which dependencies are to be shaded and including the the shaded dependencies in the package. It would limit shading to a future version of the runtime that supported such a feature, which means it would not be usable for many package authors who need to support older runtimes such as including .NET Framework.side-by-side assembly loading.
+
+### Why not AssemblyLoadContext?
+
+[`AssemblyLoadContext`](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.loader.assemblyloadcontext?view=net-5.0) allows loading side-by-side copies of the same assembly on .NET Core, Xamarin, and .NET. However, it's not supported on .NET Framework. The different copies of the assemblies must be loaded into different contexts, and it's complex to manage the contexts and marshalling across the boundaries between them when the library/app architecture has not be designed around such a boundary.
+
+### Why not ILMerge or il-repack?
+
+Both [ILMerge](https://github.com/dotnet/ILMerge) and [il-repack](https://github.com/gluck/il-repack) perform the same kind of identity-changing IL/asset manipulation required by this proposal, and we may be able to leverage them for those mechanics. However, their goal is to _merge_ assemblies, and while pack-time merging could be used to implement a producer-side shading solution, it would not offer a transition path to the consumer-side shading model into which this is intended to evolve.
+
+This proposal is also largely about how shading is specified in the project file, how it fits into the restore/build/pack pipeline, and how it affects the dependency graph. ILMerge/il-repack do not by themselves answer any of those.
+
+### How does this compare to Maven shading?
```

Comment:
The other other-ecosystem system I was thinking of while reading this was Rust's Cargo. It looks like it (effectively) _always_ shades each distinct version and provides tooling to help audit this happening: https://doc.rust-lang.org/cargo/reference/resolver.html#version-incompatibility-hazards

Reply:
Yes! I ran across that while researching this and see it as a validation of the consumer side mangling approach. It's definitely a very similar model to where I would see this going long term.

I'm not certain that such a strict interpretation of semver would work for reasoning about compatibility of NuGet packages in general, and I would like to see ways to force things to unify (e.g. to reduce size of a wasm app, when you know you're using a compatible subset of two incompatible versions) and to force things to separate (e,g. when you know there was an accidental breaking change in a minor version). But that's a problem for the future :)